File: Source/Obsequy.Web/App/Schemes/consumerPortfolioScheme.js

angular.module('consumerPortfolio-scheme', ['underscore', 'consumerMember-model'])
	.factory('ConsumerPortfolio', ['_', 'ConsumerMember', 'ConsumerPrincipal', 'ConsumerPreference', 'ConsumerRequest', 'ConsumerSchedule', function (_, ConsumerMember, ConsumerPrincipal, ConsumerPreference, ConsumerRequest, ConsumerSchedule) {
 		var ConsumerPortfolio = function (properties) {

			angular.extend(this, {

				id: undefined,
				creator: new ConsumerMember(),
				members: [],

				principal: new ConsumerPrincipal(),
				preference: new ConsumerPreference(),
				request: new ConsumerRequest(),
				schedule: new ConsumerSchedule(),

				canDelete: undefined,
				canSave: undefined,
				canSubmit: undefined,
			
				chosenOptionsCount: undefined,
				totalOptionsCount: undefined,
				percentComplete: undefined,

				// these are set by the repository
				responseIdsAccepted: [],
				responseIdsPending: [],
				responseIdsRejected: [],
				totalResponsesGenerated: undefined,
				totalResponsesAcknowledged: undefined,

				created: {},
				modified: {},
				reminded: {},

				update: function (properties) {
					properties = properties || { address: {} };

					// update properties
					this.id = properties.id;
					this.creator.update(properties.creator);

					_.each(properties.members, function (member) {
						this.members.push(new ConsumerMember(member));
					}, this);

					this.principal.update(properties.principal);
					this.preference.update(properties.preference);
					this.request.update(properties.request);
					this.schedule.update(properties.schedule);

					this.canDelete = properties.canDelete;
					this.canSave = properties.canSave;
					this.canSubmit = properties.canSubmit;

					this.chosenOptionsCount = properties.chosenOptionsCount;
					this.totalOptionsCount = properties.totalOptionsCount;
					this.percentComplete = properties.percentComplete;

					this.created = properties.created;
					this.modified = properties.modified;
					this.reminded = properties.reminded;
				},

				toJSON: function () {
					return {
						id: this.id,
						principal: this.principal.toJSON(),
						preference: this.preference.toJSON(),
						schedule: this.schedule.toJSON(),
                        request: this.request.toJSON(),
						canDelete: this.canDelete,
						canSave: this.canSave,
						canSubmit: this.canSubmit,
						chosenOptionsCount: this.chosenOptionsCount,
						totalOptionsCount: this.totalOptionsCount,
						percentComplete: this.percentComplete,
						responseIdsAccepted: this.responseIdsAccepted,
						responseIdsPending: this.responseIdsPending,
						responseIdsRejected: this.responseIdsRejected,
						totalResponsesGenerated: this.totalResponsesGenerated,
						totalResponsesAcknowledged: this.totalResponsesAcknowledged,
						created: this.created,
						modified: this.modified,
						reminded: this.reminded
					};
				}
			});

			this.update(properties);
		};

 		return ConsumerPortfolio;
	}]);

File: Source/Obsequy.Model/Scheme/Enitity/ConsumerPortfolioScheme.cs
using System.Collections.Generic;
using FluentValidation.Results;
using Obsequy.Utility;

namespace Obsequy.Model
{
	public class ConsumerPortfolioScheme
	{
		#region Properties

		public string Id { get; set; }
		public AccountStatus AccountStatus { get; set; }

		public ConsumerPrincipal Principal { get; set; }

		public ConsumerPreference Preference { get; set; }
		public ConsumerSchedule Schedule { get; set; }

		public ConsumerRequest Request { get; set; }

		public Payment Payment { get; set; }

		public List<ConsumerMember> Members { get; set; }

		public ChangeReceipt Created { get; set; }
		public ChangeReceipt Modified { get; set; }
        public ChangeReceipt Reminded { get; set; }

		public bool CanDelete { get; set; }
		public bool CanSave { get; set; }
		public bool CanSubmit { get; set; }

		public int ChosenOptionsCount { get; set; }
		public int PercentComplete { get; set; }
		public int TotalOptionsCount { get; set; }

		#endregion

		#region Constructors
		public ConsumerPortfolioScheme()
		{

		}

		public ConsumerPortfolioScheme(ConsumerPortfolio consumerPortfolio)
		{
			this.Id = consumerPortfolio.Id;
			this.AccountStatus = consumerPortfolio.AccountStatus;
			
			this.Principal = consumerPortfolio.Principal;
			this.Preference = consumerPortfolio.Preference;
			this.Schedule = consumerPortfolio.Schedule;

			this.Request = consumerPortfolio.Request;

			this.Payment = consumerPortfolio.Payment;

			this.Members = new List<ConsumerMember>();

			this.Created = consumerPortfolio.Created;
            this.Modified = consumerPortfolio.Modified;
            this.Reminded = consumerPortfolio.Reminded;

			this.CanDelete = consumerPortfolio.CanDelete;
			this.CanSave = consumerPortfolio.CanSave;
			this.CanSubmit = consumerPortfolio.CanSubmit;

			this.ChosenOptionsCount = consumerPortfolio.ChosenOptionsCount;
			this.PercentComplete = consumerPortfolio.PercentComplete;
			this.TotalOptionsCount = consumerPortfolio.TotalOptionsCount;
		}

		#endregion

		#region Methods

		#region To String
		public override string ToString()
		{
			return string.Format("{0} - {1}", this.Request.State, this.Principal.FullName);
		}
		#endregion

		#endregion
	}
}
